WIP: cert-manager & external-dns example (#3808)
garden-io/garden: docs/k8s-plugins/remote-k8s/ingress-and-dns.md

# 3. Set Up Ingress, TLS and DNS

Starting from version 0.13 Garden will no longer support the cert-manager extension (previously built-in Garden by default), this means that you will need to implement a custom solution to get it deploy into your cluster and get your certificates generated.

But don't worry! We got you covered, in this document we are going to be building a `garden configuration` that will allow us to deploy `cert-manager`, `external-dns` (generating DNS automatically on CloudFlare) and a hello-world React application that will help us to test our certificates strategy.

## About

This project deploys a small React service in a Kubernetes Cluster and exposes it to an Nginx ingress with HTTPs by using the following Kubernetes Operators:

| service       |   version  |
|---------------|------------|
| cert-manager  |  v1.11.0   |
| external-dns  |   v6.13.3  |

By using a combination of `container` and `helm` modules supported by Garden and with minimal manual intervention, we are able to rapidly deploy and provision our cluster with everything it needs for automatic TLS and DNS.

> This project was developed and battle tested using `Garden v0.12.52`.

Reviewer: ```suggestion
This project was developed and battle-tested using `Garden v0.12.52`.
```

4. Some prior knowledge in [cert-manager](https://cert-manager.io/docs/) and [ExternalDNS](https://github.com/kubernetes-sigs/external-dns) and their use cases.
5. A Docker Registry (Docker Hub, etc.) and a Kubernetes secret with the credentials of the registry.

## Project Structure Setup

First, let's start by creating our project structure and files we are going to be using across our project.

```bash
mkdir tls-and-dns && cd $_
mkdir charts
mkdir charts/cluster-issuers
mkdir frontend
touch frontend/garden.yml
touch frontend/Dockerfile
touch garden.yml
touch project.garden.yml
```

After creating those files and directories you should have the following project structure.

```bash
├── charts
│   ├── cluster-issuers <- This chart will deploy certificates and cluster-issuers that we are going to need to operate cert-manager.
├── frontend <- Will contain a React Application that we are going to use to deploy with Garden
│   ├── Dockerfile
│   ├── garden.yml
├── garden.yml <- Contains the module config for external-dns & cert-manager
├── project.garden.yml <- Contains the project configuration/env-vars.
```

> ⚠️ This project is available under our examples folder, if you have any questions or want to check if you are following the step-by-step guide correctly you can always check the source code. [examples/tls-and-dns](https://github.com/garden-io/garden/tree/main/examples)

Reviewer: ```suggestion
⚠️ For your convenience, this project is also available as a [full-code example.](https://github.com/garden-io/garden/tree/main/examples)
```

### Creating our Garden Project

First, let's make our new folder a repository.

```bash
git init
```

> And then it's time to configure the hearth of our `Garden configuration` which is the `project`.

Reviewer: ```suggestion
Now it's time to configure the heart of any Garden configuration, the `project`.
```

> Edit the `project.garden.yml` with your favorite editor and let's add the following code into it.

Reviewer: ```suggestion
Edit or create the `project.garden.yml` with your favorite editor and add the following code into it.
```

> First, the project and environment configuration. In this project we are just going to have 1 environment called `prod` so let's add the configuration for it.

Reviewer: ```suggestion
First, the project and environment configuration. In this example, we specify just one environment, `prod. 
```

Add the following content to the `project.garden.yml`

```YAML
providers:
  - name: kubernetes
    environments: [prod]
    context: your-k8s-context # Change this to your Kubernetes Context!
    namespace: ${var.DEFAULT_NAMESPACE}
    setupIngressController: nginx
    buildMode: kaniko
    kaniko:
      namespace: null
    # tlsCertificates:
    #   - name: staging-cert
    #     secretRef:
    #       name: staging-cert
    deploymentRegistry:
      hostname: "${var.registryHostname}"
      namespace: "${var.registryNamespace}"
    imagePullSecrets:
    - name: regcred
      namespace: default # Leave this in default to make configuration easier.
```

To get your context name you can simply use:

```bash
kubectl config get-contexts
kubectl config current-context
```

Now you can copy and paste your context to your `provider.context` parameter.

At this stage of the demo we are going to leave the `tlsCertificates` block commented out and we will revisit it later in a different section of this project.

> For the `deploymentRegistry` and `imagePullSecrets` we require that you already have a Docker Registry (prerequisite #5). If you already own one, please create a secret with your credentials using the following command:

Reviewer: ```suggestion
For the `deploymentRegistry` and `imagePullSecrets` we require that you already have a Docker Registry (prerequisite #5). If you already have a Docker registry on e.g. Docker Hub or Qua
```suggestion
For the `deploymentRegistry` and `imagePullSecrets` we require that you already have a Docker Registry (prerequisite #5). If you already own one, please create a secret with your credentials using the following command:
```y, please create a secret with your credentials using the following command:
```

```bash

> kubectl create secret generic regcred \

Reviewer: I would direct users to the official Kubernetes image pull secrets docs for this as users that make use of cred stores like Docker Desktop won't be able to use this method.

    --from-file=.dockerconfigjson=$HOME/.docker/config.json \
    --type=kubernetes.io/dockerconfigjson
```

If you decide to setup the secret with a different name make sure to update it in the `imagePullSecrets` configuration.

Now let's add some variables to the `project.garden.yml` that we are going to be using across our configuration in the project.

```YAML
variables:
  # Variables for Docker Registry
  registryHostname: docker.io
  registryNamespace: your-registry-namespace # Username ex. shankyweb in Dockerhub
  # Variables for Kubernetes
  DEFAULT_NAMESPACE: default # Frontend will be deployed in this namespace.
  # Cert-Manager variables
  CERT_MANAGER_INSTALL_CRDS: true
  GENERATE_PROD_CERTS: false
  GENERATE_STG_CERTS: false
  CF_DNS_PROVIDER: cloudflare # This demo only supports Cloudflare at the moment
  CF_DOMAIN: [ your-domain.com ] # Has to be an array, even if you only have one domain
  CF_UPDATE_STRATEGY: sync
  CF_EMAIL: your-email@myemail.com
  CF_PROXIED: false
```

> The variables that are required to be modified according to your environment at this moment are:

Reviewer: ```suggestion
The variables that are required to be modified according to your environment are:
```

- registryHostname: Add your registry, in this example we are going to be using Docker Hub.
- registryNamespace: Your registry namespace, in Dockerhub this will be your username.
- CF_DOMAIN: Add the DNS you will be using for this project in the form of an array.

> - CF_EMAIL: Add the email address that should get notifications from Letsencrypt/Cloudflare.

Reviewer: I would surround these variables im backticks.

  # Variables for Kubernetes
  DEFAULT_NAMESPACE: default
  # Cert-Manager variables
  CERT_MANAGER_INSTALL_CRDS: true
  GENERATE_PROD_CERTS: false
  GENERATE_STG_CERTS: false
  # External-DNS variables
  CF_DNS_PROVIDER: cloudflare
  CF_DOMAIN: [ shankyjs.com ]
  CF_UPDATE_STRATEGY: sync
  CF_EMAIL: hello@shankyjs.com
  CF_PROXIED: false

```

**Note:** Variables are going to be start to make sense after we start building our Helm Modules for `cert-manager` and `external-dns`.

### Creating our Helm Modules 🔨

Now we are going to start setting up the [Helm modules](https://docs.garden.io/reference/module-types/helm) that are going to be in charge of deploying our Helm Charts to our Kubernetes Cluster.

#### Configuring Cert-Manager module

Let's edit our `./garden.yml` file, and paste the following content, which handles the installation of the `cert-manager` operator, who is going to be in charge of creating/renewing our certificates.

For this Helm Chart we only require 1 value called `installCRDs` that is in charge of installing the [Custom Resource Definitions](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/custom-resources/) in our Kubernetes Cluster.

> Also, something to note is that we are starting to use the environment variables that we set previously at the `garden.project.yml` level.

Author: Just got rid of that part, was left over from the old document, not relevant anymore! @worldofgeese

#### Configuring cluster-issuers module

This Helm module is going to be different than the others we previously configured, because this Helm chart is intended to deploy the Custom Resources that we are going to need in order to interact with this demo, and the Helm Chart is going to be living in the same repository we are executing this demo.

At this point we are just going to add the configuration of the Helm Module for Garden, we will create the chart in the next section of this tutorial.

Add the following content to your `./garden.yml` file:

````yaml
---
# Cluster-issuers module configuration
kind: Module
type: helm
name: cluster-issuers
namespace: default
description: This module installs cluster-issuers in our Kubernetes Cluster.
chartPath: charts/cluster-issuers
dependencies:
  - cert-manager # This module depends on the cert-manager module (we need the CRDs to be installed)
values:
  cloudflare:
    email: ${var.CF_EMAIL}
    apiToken: ${local.env.CF_API_TOKEN}
    cfDomain: ${var.CF_DOMAIN[0]} # For this example, at the moment we only support one domain.
  generateStgCert: ${var.GENERATE_STG_CERTS}
  generateProdCert: ${var.GENERATE_PROD_CERTS}

> ````

Author: Unfortunately I need them as it's my only way to provide the certificates/Cluster-issuers dynamically without having to make them to apply those resources.

````yaml
{{ if .Values.generateStgCert }}
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: staging-cert
spec:
  dnsNames:
  - "react.{{ .Values.cloudflare.cfDomain }}"
  issuerRef:
    name: letsencrypt-staging
    kind: ClusterIssuer
  secretName: staging-cert
{{ end }}
---
{{ if .Values.generateProdCert }}
apiVersion: cert-manager.io/v1
kind: Certificate
metadata:
  name: production-cert
spec:
  dnsNames:
  - "react.{{ .Values.cloudflare.cfDomain }}"
  issuerRef:
    name: letsencrypt-production
    kind: ClusterIssuer
  secretName: production-cert
{{ end }}

````

#### secret.yaml

This file will contain a secret that it's going to be filled with our environment variable `CF_API_TOKEN`, we need this token as cert-manager uses it to prove that we own the domain we are using to issue certificates.

Add the following content to your secret.yaml

````yaml
apiVersion: v1
kind: Secret
metadata:
  name: cloudflare-api-token-secret
  namespace: cert-manager
type: Opaque
stringData:
  api-token: {{ .Values.cloudflare.apiToken }}

````

After creating our files, our Helm Chart is now ready to be deployed with Garden. 🪴

> 

Reviewer: I know you worked very hard on this but this takes more than 150 lines to explain a custom chart deployment. What is the value prop to a reader? Is the time the reader saves using the custom chart worth it? Is the juice worth the squeeze?

### Creating our Frontend

> To showcase our certificates we decided to use a sample react-application, the code for that React Application is hosted in this [repository](https://github.com/ShankyJS/garden-cra-demo).

Reviewer: I would re-home this inside the `examples` directory where this example lives.

You can simply clone that repository; and move the code to the folder called `frontend` in your example.

#### Cloning the existing repo

The following commands needs to be executed from your root (next to where the `project.garden.yml` file is.)

````bash
git clone git@github.com:ShankyJS/garden-cra-demo.git
mv -f garden-cra-demo frontend

> ````

Reviewer: See my earlier comment on re-homing the code sample.

This folder has the `garden.yml` and `Dockerfile` ready, so you are ready to go to the next step.

> Just to document what it's the content of the `garden.yml` to configure our React App, we are going to use the following configuration, the most important part in this configuration is the ingresses block, as it's where we are setting the hostname for the application `Ex. react.shankyjs.com`.

Reviewer: ```suggestion
The most important part of this configuration is the ingresses block, as it's where we set the hostname for the application `Ex. react.shankyjs.com`.
```

Avoid present continuous tense: when picking between "are setting" or the present tense "set", pick set. It's more direct.

      ℹ Run with --force-refresh to force a refresh of provider statuses.
   ✔ graph                     → Resolving 11 modules... → Done
   ✔ kubernetes                → Configuring... → Ready
   ℹ Run with --force-refresh to force a refresh of provider statuses.
✔ graph                     → Resolving 4 modules... → Done
✔ cluster-issuers           → Building version v-7e466ae4e3... → Done (took 0 sec)
✔ cert-manager              → Building version v-0b43e69fd6... → Done (took 0.6 sec)
✔ external-dns              → Building version v-ec953ac304... → Done (took 1.1 sec)
✔ react-app                 → Getting build status for v-8323dd9f4c... → Already built
✔ cert-manager              → Deploying version v-010a92e8da... → Done (took 37 sec)
   ℹ cert-manager              → Resources ready
✔ external-dns              → Deploying version v-11221c2d8c... → Done (took 12.9 sec)
   ℹ external-dns              → Resources ready
✔ cluster-issuers           → Deploying version v-e67a273f4c... → Already deployed
✔ frontend                  → Deploying version v-ac87f68175... → Already deployed
   Ingress: http://react.shankyjs.com

Done! ✔️
````

> Also something to note is that our website is now up, but without a certificate; because ExternalDNS already provisioned our DNS but we still haven't generated any Certificate for our frontend application.

Reviewer: We can enter on, "Our website is now up..." without altering the meaning of this sentence.

Starting from version 0.13 Garden will no longer support the cert-manager extension (previously built-in Garden by default), this means that you will need to implement your own certificate management solution if you're not using the your cloud provider's.

You'll also need to point one or more DNS entries to your cluster, and configure a TLS certificate for the hostnames
you will expose for ingress.
But don't worry! We got you covered. In this tutorial, we will use two Helm charts to deploy `cert-manager`, and `external-dns` with automatic DNS record creation on CloudFlare. We will also deploy a "Hello, World!" React application that will help us test our certificates strategy.

Templating the ingress to the application enables you to have DNS entries for every developer's namespace.
## About

First, you will make DNS CNAME entry that points to the load balancer in front of your cluster. We recommend setting a wildcard in front of the proper record, e.g. *.<environment>.<your company>.com.
This project deploys a small React service in a Kubernetes Cluster and exposes it to an Nginx ingress with HTTPs by using the following Kubernetes Operators:

> If you would like to manage TLS for development environments, we recommend using your cloud provider's certificate management service in combination with a load balancer. You can find the documentation for [AWS here](https://aws.amazon.com/premiumsupport/knowledge-center/associate-acm-certificate-alb-nlb/) and for [GCP here](https://cloud.google.com/load-balancing/docs/ssl-certificates/google-managed-certs).

Reviewer 2: I feel like this is still the main way to go if you're using Garden– the complexity of rolling your own cert-manager will, while absolutely useful for some teams, in my opinion, not be worth it for most teams.

Author: Hey @stefreak, I think I agree with this, maybe we could leave the ingress-and-dns page as a reference to all the methods we could use to have certificates with Garden with cert-manager being an important link?

Maybe this documentation I developed could live **[here](https://docs.garden.io/advanced/cert-manager-integration)** instead, replacing the deprecated cert-manager integration?

One problem that we used to see  with the #ingress-and-dns page is that the references where vague and we have users asking for how to manage the certificates with GCP for example but we currently don't have the docs to show them how that would work.

What do you think @worldofgeese ?

Reviewer 2: @ShankyJS Yes, that looks like a good location for this guide, especially as it is advanced and detailed it probably fits very well into the 'advanced' section 👍 

I would say the old guide can be improved, and you can do that if you have time for it right now, but we can also improve it later step-by-step. I would lean towards making progress, and not towards making everything perfect.

> This project was developed and battle-tested using `Garden v0.12.52`.

Reviewer 2: Not sure if this is relevant here – The example should be compatible with any coming garden version.  This version number will just grow old and confuse users who are reading this guide.

Author: I agree, this can be safely removed.
